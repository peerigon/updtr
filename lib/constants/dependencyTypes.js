"use strict";

const REGULAR = 0;
const DEV = 1;
const OPTIONAL = 2;

function fromNpm(type) {
    switch (type) {
        case "dependencies":
            return REGULAR;
        case "devDependencies":
            return DEV;
        case "optionalDependencies":
            return OPTIONAL;
        default:
            throw new Error(`Unsupported dependency type: ${ type }`);
    }
}

exports.REGULAR = REGULAR;
exports.DEV = DEV;
exports.OPTIONAL = OPTIONAL;

exports.fromNpm = fromNpm;
// yarn uses the same strings. Change this if necessary.
exports.fromYarn = fromNpm;
