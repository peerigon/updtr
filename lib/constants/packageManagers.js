"use strict";

const NPM = "npm";
const YARN = "yarn";
const SUPPORTED = [NPM, YARN];

exports.NPM = NPM;
exports.YARN = YARN;
exports.SUPPORTED = SUPPORTED;
