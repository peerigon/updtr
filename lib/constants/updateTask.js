"use strict";

const WANTED = "wanted";
const LATEST = "latest";

exports.WANTED = WANTED;
exports.LATEST = LATEST;
