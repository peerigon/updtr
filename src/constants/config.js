// The first value in arrays is the default value

export const USE_NPM = "npm";
export const USE_YARN = "yarn";
export const USE_OPTIONS = [USE_NPM, USE_YARN];
export const UPDATE_TO_LATEST = "latest";
export const UPDATE_TO_NON_BREAKING = "non-breaking";
export const UPDATE_TO_WANTED = "wanted";
export const UPDATE_TO_OPTIONS = [
    UPDATE_TO_LATEST,
    UPDATE_TO_NON_BREAKING,
    UPDATE_TO_WANTED,
];
export const SAVE_SMART = "smart";
export const SAVE_CARET = "caret";
export const SAVE_EXACT = "exact";
export const SAVE_OPTIONS = [SAVE_SMART, SAVE_CARET, SAVE_EXACT];
