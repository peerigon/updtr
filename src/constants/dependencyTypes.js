export const REGULAR = "dependencies";
export const DEV = "devDependencies";
export const OPTIONAL = "optionalDependencies";

export function fromNpm(type) {
    switch (type) {
        case "dependencies":
            return REGULAR;
        case "devDependencies":
            return DEV;
        case "optionalDependencies":
            return OPTIONAL;
        default:
            throw new Error(`Unsupported dependency type: ${ type }`);
    }
}

export {
    // yarn uses the same strings. Change this if necessary.
    fromNpm as fromYarn,
};
