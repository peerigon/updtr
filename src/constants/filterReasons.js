export const GIT = "git";
export const EXOTIC = "exotic";
export const UNSTABLE = "unstable";
export const EXCLUDED = "excluded";
export const NOT_WANTED = "not-wanted";
