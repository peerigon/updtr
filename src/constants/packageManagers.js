export const NPM = "npm";
export const YARN = "yarn";
export const SUPPORTED = [NPM, YARN];
