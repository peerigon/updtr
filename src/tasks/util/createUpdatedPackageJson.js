import updateVersionRange from "./updateVersionRange";
import { filterSuccessfulUpdates } from "./filterUpdateResults";
import { SAVE_CARET, SAVE_EXACT, SAVE_SMART } from "../../constants/config";

const dependencyTypes = [
    "dependencies",
    "devDependencies",
    "optionalDependencies",
];

function newVersionRange(updtrConfig, oldVersionRange, update) {
    switch (updtrConfig.save) {
        case SAVE_CARET:
            return "^" + update.updateTo;
        case SAVE_EXACT:
            return update.updateTo;
        case SAVE_SMART:
    }

    return updateVersionRange(oldVersionRange, update.updateTo);
}

export default function createUpdatedPackageJson(
    oldPackageJson,
    updateResults,
    updtrConfig
) {
    const newPackageJson = { ...oldPackageJson };
    const successfulUpdates = filterSuccessfulUpdates(updateResults);
    let dependenciesToSave = successfulUpdates;

    dependencyTypes
        .filter(type => oldPackageJson[type] !== undefined)
        .forEach(type => {
            const dependencies = oldPackageJson[type];
            const newDependencies = {};

            Object.keys(dependencies).forEach(moduleName => {
                const update = successfulUpdates.find(
                    ({ name }) => name === moduleName
                );
                const oldVersionRange = dependencies[moduleName];

                newDependencies[moduleName] = update === undefined ?
                    oldVersionRange :
                    newVersionRange(updtrConfig, oldVersionRange, update);

                dependenciesToSave = dependenciesToSave.filter(
                    ({ name }) => name !== moduleName
                );
            });

            newPackageJson[type] = newDependencies;
        });

    if (dependenciesToSave.length > 0) {
        const dependencies = newPackageJson.dependencies || {};

        dependenciesToSave.forEach(update => {
            dependencies[update.name] = update.updateTo;
        });

        newPackageJson.dependencies = dependencies;
    }

    return newPackageJson;
}
