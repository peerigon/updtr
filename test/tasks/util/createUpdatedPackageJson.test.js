import clone from "clone";
import { SAVE_CARET, SAVE_EXACT } from "../../../src/constants/config";
import createUpdatedPackageJson
    from "../../../src/tasks/util/createUpdatedPackageJson";
import {
    module1ToLatestSuccess,
    module2ToLatestSuccess,
    module1ToLatestFail,
    module1ToNonBreakingFail,
    module2ToLatestFail,
} from "../../fixtures/updateResults";
import FakeUpdtr from "../../helpers/FakeUpdtr";

// Freezes the essential parts of the package.json to detect changes on the input object
function cloneAndFreeze(packageJson) {
    const cloned = clone(packageJson);

    Object.freeze(cloned.dependencies);
    Object.freeze(cloned.devDependencies);
    Object.freeze(cloned.optionalDependencies);

    return Object.freeze(cloned);
}

describe("createUpdatedPackageJson()", () => {
    it("should not create new fields by default", () => {
        const updateResults = [];
        const newPackageJson = createUpdatedPackageJson(
            Object.freeze({}),
            updateResults,
            FakeUpdtr.baseConfig
        );

        expect(newPackageJson).toEqual({});
    });
    it("should not create new fields for failed update results", () => {
        const updateResults = [module1ToLatestFail, module2ToLatestFail];
        const newPackageJson = createUpdatedPackageJson(
            Object.freeze({}),
            updateResults,
            FakeUpdtr.baseConfig
        );

        expect(newPackageJson).toEqual({});
    });
    describe("dependencies", () => {
        it("should write the versions of the successful updates to the package json", () => {
            const updateResults = [
                module1ToLatestSuccess,
                module2ToLatestSuccess,
                {
                    ...module1ToLatestFail,
                    name: "updtr-test-module-3",
                },
                {
                    ...module1ToNonBreakingFail,
                    name: "updtr-test-module-4",
                },
                {
                    ...module1ToLatestSuccess,
                    name: "updtr-test-module-5",
                },
            ];
            const newPackageJson = createUpdatedPackageJson(
                cloneAndFreeze({
                    dependencies: {
                        "updtr-test-module-1": "1.0.0",
                        "dep-1": "1.0.0",
                        "updtr-test-module-2": "1.0.0",
                        "dep-2": "1.0.0",
                        "updtr-test-module-3": "1.0.0",
                        "dep-3": "1.0.0",
                    },
                }),
                updateResults,
                FakeUpdtr.baseConfig
            );

            expect(newPackageJson).toMatchSnapshot(
                "dependencies > only successful updates"
            );
        });
        describe(`when the save option is "${ SAVE_EXACT }"`, () => {
            it("should save the exact version", () => {
                const updateResults = [module1ToLatestSuccess];
                const newPackageJson = createUpdatedPackageJson(
                    cloneAndFreeze({
                        dependencies: {
                            "updtr-test-module-1": "^1.0.0",
                        },
                    }),
                    updateResults,
                    { ...FakeUpdtr.baseConfig, save: SAVE_EXACT }
                );

                expect(newPackageJson.dependencies["updtr-test-module-1"]).toBe(
                    module1ToLatestSuccess.updateTo
                );
            });
        });
        describe(`when the save option is "${ SAVE_CARET }"`, () => {
            it("should save the exact version", () => {
                const updateResults = [module1ToLatestSuccess];
                const newPackageJson = createUpdatedPackageJson(
                    cloneAndFreeze({
                        dependencies: {
                            "updtr-test-module-1": "1.0.0",
                        },
                    }),
                    updateResults,
                    { ...FakeUpdtr.baseConfig, save: SAVE_CARET }
                );

                expect(newPackageJson.dependencies["updtr-test-module-1"]).toBe(
                    "^" + module1ToLatestSuccess.updateTo
                );
            });
        });
    });
    describe("devDependencies", () => {
        it("should write the versions of the successful updates to the package json", () => {
            const updateResults = [
                module1ToLatestSuccess,
                module2ToLatestSuccess,
                {
                    ...module1ToLatestFail,
                    name: "updtr-test-module-3",
                },
                {
                    ...module1ToNonBreakingFail,
                    name: "updtr-test-module-4",
                },
            ];
            const newPackageJson = createUpdatedPackageJson(
                cloneAndFreeze({
                    devDependencies: {
                        "updtr-test-module-1": "1.0.0",
                        "dep-1": "1.0.0",
                        "updtr-test-module-2": "1.0.0",
                        "dep-2": "1.0.0",
                        "updtr-test-module-3": "1.0.0",
                        "dep-3": "1.0.0",
                    },
                }),
                updateResults,
                FakeUpdtr.baseConfig
            );

            expect(newPackageJson).toMatchSnapshot(
                "devDependencies > only successful updates"
            );
        });
        describe(`when the save option is "${ SAVE_EXACT }"`, () => {
            it("should save the exact version", () => {
                const updateResults = [module1ToLatestSuccess];
                const newPackageJson = createUpdatedPackageJson(
                    cloneAndFreeze({
                        devDependencies: {
                            "updtr-test-module-1": "^1.0.0",
                        },
                    }),
                    updateResults,
                    { ...FakeUpdtr.baseConfig, save: SAVE_EXACT }
                );

                expect(
                    newPackageJson.devDependencies["updtr-test-module-1"]
                ).toBe(module1ToLatestSuccess.updateTo);
            });
        });
        describe(`when the save option is "${ SAVE_CARET }"`, () => {
            it("should save the exact version", () => {
                const updateResults = [module1ToLatestSuccess];
                const newPackageJson = createUpdatedPackageJson(
                    cloneAndFreeze({
                        dependencies: {
                            "updtr-test-module-1": "1.0.0",
                        },
                    }),
                    updateResults,
                    { ...FakeUpdtr.baseConfig, save: SAVE_CARET }
                );

                expect(newPackageJson.dependencies["updtr-test-module-1"]).toBe(
                    "^" + module1ToLatestSuccess.updateTo
                );
            });
        });
    });
    describe("optionalDependencies", () => {
        it("should write the versions of the successful updates to the package json", () => {
            const updateResults = [
                module1ToLatestSuccess,
                module2ToLatestSuccess,
                {
                    ...module1ToLatestFail,
                    name: "updtr-test-module-3",
                },
                {
                    ...module1ToNonBreakingFail,
                    name: "updtr-test-module-4",
                },
            ];
            const newPackageJson = createUpdatedPackageJson(
                cloneAndFreeze({
                    optionalDependencies: {
                        "updtr-test-module-1": "1.0.0",
                        "dep-1": "1.0.0",
                        "updtr-test-module-2": "1.0.0",
                        "dep-2": "1.0.0",
                        "updtr-test-module-3": "1.0.0",
                        "dep-3": "1.0.0",
                    },
                }),
                updateResults,
                FakeUpdtr.baseConfig
            );

            expect(newPackageJson).toMatchSnapshot(
                "optionalDependencies > only successful updates"
            );
        });
        describe(`when the save option is "${ SAVE_EXACT }"`, () => {
            it("should save the exact version", () => {
                const updateResults = [module1ToLatestSuccess];
                const newPackageJson = createUpdatedPackageJson(
                    cloneAndFreeze({
                        optionalDependencies: {
                            "updtr-test-module-1": "^1.0.0",
                        },
                    }),
                    updateResults,
                    { ...FakeUpdtr.baseConfig, save: SAVE_EXACT }
                );

                expect(
                    newPackageJson.optionalDependencies["updtr-test-module-1"]
                ).toBe(module1ToLatestSuccess.updateTo);
            });
        });
        describe(`when the save option is "${ SAVE_CARET }"`, () => {
            it("should save the exact version", () => {
                const updateResults = [module1ToLatestSuccess];
                const newPackageJson = createUpdatedPackageJson(
                    cloneAndFreeze({
                        dependencies: {
                            "updtr-test-module-1": "1.0.0",
                        },
                    }),
                    updateResults,
                    { ...FakeUpdtr.baseConfig, save: SAVE_CARET }
                );

                expect(newPackageJson.dependencies["updtr-test-module-1"]).toBe(
                    "^" + module1ToLatestSuccess.updateTo
                );
            });
        });
    });
    // This is usually an error, but we don't care
    describe("when a module is listed in more than one dependency type", () => {
        it("should update all dependency occurrences", () => {
            const updateResults = [module1ToLatestSuccess];
            const newPackageJson = createUpdatedPackageJson(
                cloneAndFreeze({
                    dependencies: {
                        "updtr-test-module-1": "1.0.0",
                    },
                    devDependencies: {
                        "updtr-test-module-1": "^1.0.0",
                    },
                    optionalDependencies: {
                        "updtr-test-module-1": "1.x.x",
                    },
                }),
                updateResults,
                FakeUpdtr.baseConfig
            );

            expect(newPackageJson).toMatchSnapshot(
                "more than one dependency type"
            );
        });
    });
    describe("when the updated modules are not listed as any dependency type", () => {
        it("should save the modules as regular dependency", () => {
            const updateResults = [
                module1ToLatestSuccess,
                module2ToLatestSuccess,
            ];
            const newPackageJson = createUpdatedPackageJson(
                Object.freeze({}),
                updateResults,
                FakeUpdtr.baseConfig
            );

            expect(newPackageJson).toMatchSnapshot(
                "not listed as any dependency type"
            );
        });
    });
});
