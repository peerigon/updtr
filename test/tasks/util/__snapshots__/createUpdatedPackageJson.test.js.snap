// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dependencies > only successful updates 1`] = `
Object {
  "dependencies": Object {
    "dep-1": "1.0.0",
    "dep-2": "1.0.0",
    "dep-3": "1.0.0",
    "updtr-test-module-1": "2.0.0",
    "updtr-test-module-2": "2.1.1",
    "updtr-test-module-3": "1.0.0",
    "updtr-test-module-5": "2.0.0",
  },
}
`;

exports[`devDependencies > only successful updates 1`] = `
Object {
  "devDependencies": Object {
    "dep-1": "1.0.0",
    "dep-2": "1.0.0",
    "dep-3": "1.0.0",
    "updtr-test-module-1": "2.0.0",
    "updtr-test-module-2": "2.1.1",
    "updtr-test-module-3": "1.0.0",
  },
}
`;

exports[`more than one dependency type 1`] = `
Object {
  "dependencies": Object {
    "updtr-test-module-1": "2.0.0",
  },
  "devDependencies": Object {
    "updtr-test-module-1": "^2.0.0",
  },
  "optionalDependencies": Object {
    "updtr-test-module-1": "2.x.x",
  },
}
`;

exports[`not listed as any dependency type 1`] = `
Object {
  "dependencies": Object {
    "updtr-test-module-1": "2.0.0",
    "updtr-test-module-2": "2.1.1",
  },
}
`;

exports[`optionalDependencies > only successful updates 1`] = `
Object {
  "optionalDependencies": Object {
    "dep-1": "1.0.0",
    "dep-2": "1.0.0",
    "dep-3": "1.0.0",
    "updtr-test-module-1": "2.0.0",
    "updtr-test-module-2": "2.1.1",
    "updtr-test-module-3": "1.0.0",
  },
}
`;
